sway-core/src/ir_generation/const_eval.rs

```diff
@@ -730,82 +730,121 @@ fn const_eval_intrinsic(
+                            ty,
+                            value: ConstantValue::Uint(sum),
+                        })),
+                        None => Err(ConstEvalError::CannotBeEvaluatedToConst {
+                            span: intrinsic.span.clone(),
+                        }),
+                    }
+                }
+                (U256(arg1), U256(arg2)) => {
+                    let result = match intrinsic.kind {
+                        Intrinsic::Add => arg1.checked_add(arg2),
+                        Intrinsic::Sub => arg1.checked_sub(arg2),
+                        Intrinsic::Mul => arg1.checked_mul(arg2),
+                        Intrinsic::Div => arg1.checked_div(arg2),
+                        Intrinsic::And => Some(arg1.bitand(arg2)),
+                        Intrinsic::Or => Some(arg1.bitor(arg2)),
+                        Intrinsic::Xor => Some(arg1.bitxor(arg2)),
+                        Intrinsic::Mod => Some(arg1.rem(arg2)),
+                        _ => unreachable!(),
+                    };
+
+                    match result {
+                        Some(sum) => Ok(Some(Constant {
+                            ty,
+                            value: ConstantValue::U256(sum),
+                        })),
+                        None => Err(ConstEvalError::CannotBeEvaluatedToConst {
+                            span: intrinsic.span.clone(),
+                        }),
+                    }
+                }
+                _ => {
+                    panic!("Type checker allowed incorrect args to binary op");
+                }
             }
         }
-        sway_ast::Intrinsic::Lsh | sway_ast::Intrinsic::Rsh => {
+        Intrinsic::Lsh | Intrinsic::Rsh => {
             let ty = args[0].ty;
-            assert!(
-                args.len() == 2
-                    && ty.is_uint(lookup.context)
-                    && args[1].ty.is_uint64(lookup.context)
-            );
-
-            let (ConstantValue::Uint(arg1), ConstantValue::Uint(ref arg2)) =
-                (&args[0].value, &args[1].value)
-            else {
-                panic!("Type checker allowed incorrect args to binary op");
-            };
-
-            let result = match intrinsic.kind {
-                Intrinsic::Lsh => u32::try_from(*arg2)
-                    .ok()
-                    .and_then(|arg2| arg1.checked_shl(arg2)),
-                Intrinsic::Rsh => u32::try_from(*arg2)
-                    .ok()
-                    .and_then(|arg2| arg1.checked_shr(arg2)),
-                _ => unreachable!(),
-            };
-
-            match result {
-                Some(sum) => Ok(Some(Constant {
-                    ty,
-                    value: ConstantValue::Uint(sum),
-                })),
-                None => Err(ConstEvalError::CannotBeEvaluatedToConst {
-                    span: intrinsic.span.clone(),
-                }),
+            assert!(args.len() == 2 && args[1].ty.is_uint64(lookup.context));
```

Comment:
This needs type checking of the first argument. It's more relaxed because it's not longer just classic uints, but it's still not everything. You can't shift a struct.

Reply:
The type checking is actually done before const_eval is called, that is why we panic here in some cases. But I returned the asserts because they do make sense.

Reply:
Oh yeah I didn't mean literal type checking but more complete asserts.